CJA 16 UI Friction Update

frankiebarrios opened the pull request:
## What It Does
This pull request is set to complete UI Friction updates.

This work is taking place to complete UI Friction updates.

**Implementation:**
- Added icons to forms.

- Fixed Webpack/ ‘fix-polymer-imports’ Loader issues.

- Removed buttons from forms.

- Separated <user-list> from other components.

- Changed UI to render in a horizontal/ row vs in column view.

- Added form validation.

- Moved deleteAllUsers() trigger.

- CSS Refactor/ Clean Up

## How To Test
- Install dependencies using instruction in README.
- Start server using yarn start or build using yarn build.

## Notes/Caveats

None
## Checklist

Under penalty of public shaming, I avow that I:
**Many items in list N/A** 
- [X] Reviewed the diff to look for typos, whitespace errors, and console.log() statements
- [X] Verified that there are no compiler or linter errors or warnings
- [X] Verified the build in production(optimized) mode
- [X] Updated the docs, if necessary
- [X] Included the appropriate labels
- [X] Referenced applicable issues (i.e. Closes #XXXX, Fixes #XXXX)

Browsers tested:
**Many items in list N/A** 
- [X] Chrome
- [X] Firefox
- [ ] Safari
- [ ] Edge
- [ ] Internet Explorer

## Dependencies
1. Install Node and npm
2. Install Yarn
3. Install Ux-Lint

Commit:
- Resolved issue with user cards not being deleted when deleteAllUsers() fired.
- Added deleteUser() trigger to user cards.
- Removed some unused code.

Commit:
- Resolved compiling issue, re-installed Firebase.

Commit:
no message

Commit:
Working on getting icons in and userlist populated upon load.

Commit:
- Resolved issue with users not populating UI cards upon app load.

Commit:
Refactored loadUsers()

Commit:
Working on Webpack Issue

Commit:
Most recent Webpack changes

Commit:
- Most current work on getting icons into app.

Commit:
- Resolved issue with jha-design-components. Clean up still needed.

Commit:
Check in to switch branches

Commit:
- Resolved issue with fix-imports-loader.
- Still having issues with icon rendering but files are being converted successfully through loader now.

Commit:
- Refactored fix-polymer-imports loader.

Commit:
- Resolved issues with loader/ Webpack Config, icons now rendering successfully.

Commit:
Clean Up

Commit:
- UI Friction changes.
- Removed a lot of unused styling.
- General CSS Clean Up

Commit:
- Further UI improvements.

Commit:
Clean Up

kddrake commented on src/api/firebaseAPI.js:
```diff
@@ -0,0 +1,95 @@
+    }
+  }
+
+  syncData (user) {
```
I'm not understanding how this is syncing, it looks like it's just deleting?

frankiebarrios replied:
It's adding the user to the db using createUser() and deleting the old record it replaces using deleteUser().

kddrake commented on src/components/create-user.html:
```diff
@@ -115,18 +121,83 @@ <h1 class="div-center">Create User</h1>
+        }
+      }
+
+      testFunction() {
```
remove this

kddrake commented on tools/fix-polymer-imports.js:
```diff
@@ -0,0 +1,18 @@
+const PolymerRegex2 = /class extends Polymer\.Element/;
+
+module.exports = (content) => {
+  console.log('Content before if:  ----------------------------------------------------------------,', content);
```
can we remove these console logs?

kddrake commented on testWPconfig.js:
```diff
@@ -0,0 +1,78 @@
+const webpack = require('webpack');
```
if the real webpack config is good, can we get rid of this one?

kddrake commented on src/components/user-profile.html:
```diff
@@ -77,27 +34,41 @@
+        fill:#6495ED;
+      }
+
+      .iconContainer {
```
rather than setting a class, just style the icon element itself
```suggestion
      jha-icon-circle-plus-outline {
```

frankiebarrios replied:
This class is just a container that I'm using to position my icons in my UI, not the icons themselves? Not sure what you mean here? 
Also I removed the .icon class that I had coloring my icons and just tried
 icon { fill: #6495ED }; but it didn't take? Do the jha-icons read as icon elements for CSS?

kddrake replied:
@frankielb they should be read like any other element in the template (here's an example from Reports).

I guess I was thinking a container shouldn't be necessary for the icon but I'm not entirely sure why at this point.

kddrake commented on src/components/create-user.html:
```diff
@@ -115,18 +121,83 @@ <h1 class="div-center">Create User</h1>
         this.dispatchEvent(new CustomEvent('updateUserList', { bubbles: true, composed: true }));
       }
+
+      validateInput() {
```
this could probably be split out into multiple functions, and probably should be.

frankiebarrios replied:
Yeah I agree for sure. Will do.

kddrake commented on src/components/create-user.html:
```diff
@@ -115,18 +121,83 @@ <h1 class="div-center">Create User</h1>
-      updateUserList() {
-        this.createUser();
+      deleteAllUsers() {
+        this.storage.deleteAllUsers();       
```
remove the trailing whitespace

kddrake reviewed (changes requested):
- Need to remove `package-lock.json` from the PR
- `localStorage.js` needs a bunch of ending `;`s

kddrake reviewed (changes requested):
there are some merge conflicts that'll need to be fixed

Commit:
- Removed console logs from file.

Commit:
- Removed test webpack config file.

Commit:
- Removed whitespace.
- Added semicolons back to file. Formatter deleted them.

Commit:
- Removed package-lock.json.
- Used inline styling vs class for icon color in user-profile.

Commit:
- Cleaned up unused code.
- Refactored validateInput() and changed call.

frankiebarrios committed:
Merge branch 'master' into CJA-16-UI-Friction-Update

kddrake commented on src/components/create-user.html:
```diff
@@ -125,14 +155,73 @@ <h2 class="div-center">Create New User</h2>
+      validateFirstName() {
+        const phoneRegex = /^\d{10}$/;
+        if (!this.user.firstName) {
+          this.firstNameError = 'Must Enter A First Name';
```
does this ever change? if not, you might as well just assign it up in `static get properties()`. Same with the rest of the `this.*Error` properties.

kddrake commented on src/components/create-user.html:
```diff
@@ -125,14 +155,73 @@ <h2 class="div-center">Create New User</h2>
+        if (!this.user.firstName) {
+          this.firstNameError = 'Must Enter A First Name';
+          this.hasError = true;
+          this.$.firstName.setAttribute('class', 'hasError');
```
we try to keep as much direct DOM manipulation out of the JS as possible. You should be able to bind some function to firstName input's `class` and `placeholder` attributes just like anything else.

kddrake commented on src/components/user-profile.html:
```diff
@@ -78,22 +34,40 @@
+        </template>
+        <template is="dom-if" if="[[profileExpanded]]">
+          <div class="iconContainer">
+            <jha-icon-circle-minus-outline on-click="expandCollapseCard" style="fill:#6495ED;"></jha-icon-circle-minus-outline>
```
you're not going to want the inline styling, pull it up into your style tag. You'll want to do the same with the rest of this file, too.

frankiebarrios replied:
Removed the .icon class that I had coloring my icons and the inline styling and just tried
icon { fill: #6495ED }; in the <style> tags but it didn't take? Do the jha-icons read as icon elements for CSS?

frankiebarrios replied:
Done

Commit:
- Refactored validation functions, props, and attributes on input elements.

Commit:
Merge branch 'CJA-16-UI-Friction-Update' of github.com:frankielb/frankie-barrios-ux-iop into CJA-16-UI-Friction-Update

Commit:
- Style Fix

Commit:
- Changed styling method for icons.

kddrake commented on src/components/user-profile.html:
```diff
@@ -78,22 +34,44 @@
+        justify-content: flex-end;
+      }
+
+      .iconContainer2 {
```
this is a pretty terrible naming convention. Should really try to come up with a more descriptive name, like this one `profileIcons` and the other `visibilityIcons` would be better than just `1` and `2`

frankiebarrios replied:
Done

kddrake reviewed (changes requested):
- all of your user profile cards in your user list expands when one card expands, I wouldn't expect this to happen.
- saving a user profile doesn't clear the form, which I would expect to happen.
- there's an error that's thrown when deleting all users or a single user
```
Error Pulling In Users: Database Empty  TypeError: Cannot convert undefined or null to object
```

The first issue I'd be alright with letting slide, but the other two should be fixed.

Commit:
- Removed error you would get upon trying to pull in users from an empty database.
- Renamed icon container classes.
- Added logic to clear form when user created.

kddrake commented on src/components/create-user.html:
```diff
@@ -190,6 +190,7 @@ <h5>Create</h5>
         this.validateInput();
         if(!this.hasError) {
           this.storage.createUser(this.user);
+          this.$.createUserForm.reset();
```
again, avoid the direct dom manipulation. I'd rather you reset your `user` object over this.

frankiebarrios replied:
Quick catch lol

frankiebarrios replied:
Done

Commit:
- Removed direct dom manipulation when new user created.

Commit:
- Clean up

kddrake commented on src/components/create-user.html:
```diff
@@ -190,7 +190,7 @@ <h5>Create</h5>
         if(!this.hasError) {
           this.storage.createUser(this.user);
-          this.$.createUserForm.reset();
+          this.user = '';
```
`user` is an object, not a string

Commit:
- Corrected type used in this.user reset.
